### PR TITLE
Gain some perf

### PR DESCRIPTION
**Pull request by pakrym**
Goes from 133k to 154k

Check bytes left before reading or IIS would try to read from network.

Also lazy features, a bit less tasks.

### CODE REVIEW

File: src/RequestHandler/managedexports.cxx [hunk 1]
```diff
@@ -206,18 +206,26 @@ http_read_request_bytes(
...
-
-    hr = pHttpRequest->ReadEntityBody(
-        pvBuffer,
-        dwCbBuffer,
-        fAsync,
-        pdwBytesReceived,
-        pfCompletionPending);
 
-    if (hr == HRESULT_FROM_WIN32(ERROR_HANDLE_EOF))
+    if (pHttpRequest->GetRemainingEntityBytes() > 0)
```
Comment (jkotalik) [inline src/RequestHandler/managedexports.cxx]: How does this affect read performance? Is GetRemainingEntityBytes inexpensive when there is data available? Doing this on every read could be costly.

Comment (pakrym) [reply src/RequestHandler/managedexports.cxx]: Shouldn't affect read performance it returns 0 for GET requests and some value for Content-Lenght, for everything else `INFINITE` is returned.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 2]
```diff
@@ -122,11 +144,12 @@ private static NativeMethods.REQUEST_NOTIFICATION_STATUS HandleRequest(IntPtr pI
             var server = (IISHttpServer)GCHandle.FromIntPtr(pvRequestContext).Target;
             Interlocked.Increment(ref server._outstandingRequests);
 
-            var context = server._iisContextFactory.CreateHttpContext(pInProcessHandler);
-
-            var task = Task.Run(() => context.ProcessRequestAsync());
-
-            task.ContinueWith((t, state) => CompleteRequest((IISHttpContext)state, t), context);
+            _ = Task.Run(
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: @davidfowl did we want to put this back on the IIS Request thread?

File: src/RequestHandler/managedexports.cxx [hunk 3]
```diff
@@ -206,18 +206,26 @@ http_read_request_bytes(
...
+            pvBuffer,
+            dwCbBuffer,
+            fAsync,
+            pdwBytesReceived,
+            pfCompletionPending);
+
+        if (hr == HRESULT_FROM_WIN32(ERROR_HANDLE_EOF))
+        {
+            // We reached the end of the data
+            hr = S_OK;
```
Comment (jkotalik) [inline src/RequestHandler/managedexports.cxx]: Set hr = S_OK initially so we don't need the else. Also, don't we need to set pdwBytesReceived to 0 and pfCompletionPending to False?

Comment (pakrym) [reply src/RequestHandler/managedexports.cxx]: Done

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 2 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: No closures...

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: It looks better and taking ctor off iis thread gains ~4k. I say lets take it and iterate more later.

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: Get rid of the closure and use QUWI?

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: There is a task created for state machine anyway

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: My bad, you are right. Will change.

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: But you can avoid the per request allocation....

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: Also, cross compile for nca2.1 and compile out inproc from netstandard

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 4]
```diff
@@ -144,16 +144,25 @@ private static NativeMethods.REQUEST_NOTIFICATION_STATUS HandleRequest(IntPtr pI
...
-                });
-
+#if NETCOREAPP2_1
+            ThreadPool.QueueUserWorkItem<(IISHttpServer, IntPtr)>(
+                state => _ = HandleRequest(state.Item1, state.Item2),
+                (server, pInProcessHandler),
+                preferLocal: false);
+#else
+            ThreadPool.QueueUserWorkItem(
+                state => _ = HandleRequest(server, pInProcessHandler));
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: But why?

**Comment by pakrym** [comment]
@davidfowl any other concerns?

**Pull request merged**